spec/registry.md

```diff
@@ -381,6 +381,70 @@ together with the resolved options' values.
 
 The _function_ `:integer` performs selection as described in [Number Selection](#number-selection) below.
 
+### The `:math` function
+
+The function `:math` is a selector and formatter for matching or formatting
+numeric values to which a mathematical operation has been applied.
+
+> This function is useful for selection and formatting of values that
+> differ from the input value by a specified amount.
+> For example, it can be used in a message such as this:
+> ```
+> .input {$like_count :integer}
+> .local $others_count = {$like_count :math subtract=1}
+> .match $like_count $others_count
+> 0 * {{Your post has no likes.}}
+> 1 * {{{$name} liked your post.}}
+> * 1 {{{$name} and one other person liked your post.}}
+> * * {{{$name} and {$others_count} other people liked your post.}}
+> ```
+
+#### Operands
+
+The function `:math` requires a [Number Operand](#number-operands) as its _operand_.
+
+#### Options
+
+The options on `:math` are exclusive with each other,
+and exactly one option is always required.
+The options do not have default values.
+
+The following options and their values are
+required in the default registry to be available on the function `:math`:
+- `add`
+  - ([digit size option](#digit-size-options))
+- `subtract`
+  - ([digit size option](#digit-size-options))
+
+If no options or more than one option is set,
+or if an _option_ value is not a [digit size option](#digit-size-options),
+a _Bad Option_ error is emitted
+and a _fallback value_ used as the _resolved value_ of the _expression_.
+
+#### Resolved Value
+
+The _resolved value_ of an _expression_ with a `:math` _function_
+contains the implementation-defined numeric value
+of the _operand_ of the annotated _expression_.
+
+If the `add` option is set,
+the numeric value of the _resolved value_ is formed by incrementing
+the numeric value of the _operand_ by the integer value of the digit size option value.
+
+If the `subtract` option is set,
+the numeric value of the _resolved value_ is formed by decrementing
+the numeric value of the _operand_ by the integer value of the digit size option value.
+
+If the _operand_ of the _expression_ is an implementation-defined numeric type,
+such as the _resolved value_ of an _expression_ with a `:number` or `:integer` _annotation_,
+it can include option values.
+These are included in the resolved option values of the _expression_.
+The `:math` _options_ are not included in the resolved option values.
+
+#### Selection
+
+The _function_ `:math` performs selection as described in [Number Selection](#number-selection) below.
+
```

Comment:
I see some significant problems in these formulations.

>The _resolved value_ of an _expression_ with a `:number` _function_
contains an implementation-defined numerical value
of the _operand_ of the annotated _expression_,
together with the resolved options' values.

If that is all the resolved value is, then there is no difference in resolved value between the following two.
```
.local $x1 = {$x :integer}
.local $x2 = {$x :number}
```

But of course there is a big difference: the resolved value of $x1 contains a :integer annotation and the resolved value of $x2 contains a :number annotation, as in:
```
such as the _resolved value_ of an _expression_ with a `:number` or `:integer` _annotation_,
```

And it has to be that way, otherwise no .local setting later on could tell the difference between them.

So I think L142 of formatting.md is insufficient:

> ```ts
> interface MessageValue {
>   formatToString(): string
>   formatToX(): X // where X is an implementation-defined type
>   getValue(): unknown
>   resolvedOptions(): { [key: string]: MessageValue }
>   selectKeys(keys: string[]): string[]
> }
> ```

It also needs do have something like: getFunctionAnnotation();

In shorthand, the resolved value needs to logically contain a triple {datatype, functionAnnotation, optionMap}. Otherwise, a subsequent .local statement using a resolved value wouldn't have enough information to know how to compose.

So let's look at the following:
```
The _operand_ of a number function is either an implementation-defined type or
a literal whose contents match the `number-literal` production in the [ABNF](/spec/message.abnf).
All other values produce a _Bad Operand_ error.
```

But the operand can also be a resolved value, which is neither a literal nor just an implementation datatype; it has additional information.

And in multiple cases in this file, we have wording like:

>Resolved Value
>The resolved value of an expression with a :number function contains an implementation-defined numerical value of the operand of the annotated expression, together with the resolved options' values.

Now, if somewhere we say that the resolved value of any expression with a function contains that function as an annotation, then we could get away with this shorthand (although we also need to fix the wording for options). 

Otherwise, we need to say:
>The resolved value of an expression with a :number function contains an implementation-defined numerical value of the operand of the annotated expression, together with **the :number function annotation, and** the **options with their respective** resolved option~'s~ values.

Reply:
I don't get it. 

The composition model is that functions can expose specific options, but the identity of the function itself is not passed. There is no difference between the resolved values of `{$x :number}` and `{$x :integer}` with regard to the receiving function: it's just a numeric type. A subsequent local declaration doesn't need to know which annotation was previously applied: the preceding function handler modified the resolved value and/or passed the relevant options to communicate what happened.

It has to be that way, because otherwise each function would need to understand what all of the other functions (including namespaced ones) "mean".

If `{$x :math subtract=2}` doesn't return `$x - 2` as its resolved value, `:number` isn't going to know what to do.

There are places where I have actively resisted "doing harm" to the resolved value based on the annotation. That's an argument we can have. Mostly I'm concerned that we don't strip information from the resolved value because of some formatting instructions:

```
.input {$now :date style=medium}
.local $time = {$now :time style=short}
{{"Today is {$now} and it is {$time}" only works if :date doesn't remove the time.
    I assume that is okay because immutability says that if $now is a Date object or ZonedDateTime
    there is no need to convert it to just a ZonedDate and make the time midnight.}}
```

We could change `:integer` to make the resolved value the floor value of the operand or to make it the implementation-defined integer-type casting of the operand. That doesn't break immutability because the resolved value is only exposed in a declaration. I think that's the better course than what you're proposing.

Reply:
I think I conflated too many thoughts in my message. So put on hold the notion of the function annotation (not needed for release) and I focus on just 2 issues.

1. **The operand for a function may also be a resolved value** (which includes an optionMap). That is not the same as an implementation defined type (eg Date); a resolved value can **have an** implementation-defined type. Thus 

OLD
> The _operand_ of a number function is either an implementation-defined type or
a literal whose contents match the `number-literal` production in the [ABNF](/spec/message.abnf).
All other values produce a _Bad Operand_ error.

needs to be changed to:

NEW
> The _operand_ of a number function is either an implementation-defined type or
a literal whose contents match the `number-literal` production in the [ABNF](/spec/message.abnf),
**or a resolved value.**
All other values produce a _Bad Operand_ error.

(with similar changes for other definitions of _operand_ of an X function.)

2. **The expression "together with the resolved options' values." is not accurate.** For {|3| :number roundingIncrement=5 trailingZeroDisplay=auto}, that would imply a logical result of 

`{operand=3, resolvedOptionValues=[5, auto]} 
`
rather than the logical result of

`{operand=3, resolvedOptionMap={roundingIncrement=5, trailingZeroDisplay=auto}} 
`
So we need to have:

OLD
> ... value of the operand of the annotated expression, together with the resolved options' values.

changed to something like:

NEW
> ... value of the operand of the annotated expression, together **with a map from options to their resolved** values

Reply:
The problem with the first item is that the resolved value might not be of the correct implementation-defined type. The current text accounts for this by not mentioning. We could say instead:

> The _operand_ of a number _function_ is either an implementation-defined type or
> a _literal_ whose contents match the `number-literal` production in the [ABNF](message.abnf)
> (either of which can be the _resolved value_ of a previous _declaration_).
> All other values produce a _Bad Operand_ error.

The second suggestion we already discussed. Saying "map" was thought too prescriptive (ditto "array", "list", etc.). I can see how one might read "options' values" to mean **only** the values and not the k-v pairing. We might say:

> ... value of the _operand_ of the annotated _expression_, together with _options_ and their resolved values.

Reply:
From Addison

> The problem with the first item is that the resolved value might not be of the correct implementation-defined type. The current text accounts for this by not mentioning. We could say instead:
> 
> > The _operand_ of a number _function_ is either an implementation-defined type or
> > a _literal_ whose contents match the `number-literal` production in the [ABNF](message.abnf)
> > (either of which can be the _resolved value_ of a previous _declaration_).
> > All other values produce a _Bad Operand_ error.

Wouldn't quite work. Again, we have defined the resolved value of {|3| :number style=percent} to be **NOT** the same as |3| — since the resolved value includes options. I think what we could say:

> > (either of which can be from the _resolved value_ of a previous _declaration_).

I think that is at least headed in the right direction, though not as correct as 

> or a resolved value with those as an operand.

> 
> The second suggestion we already discussed. Saying "map" was thought too prescriptive (ditto "array", "list", etc.). I can see how one might read "options' values" to mean **only** the values and not the k-v pairing. We might say:
> 
> > ... value of the _operand_ of the annotated _expression_, together with _options_ and their resolved values.

That sounds perfect.

Reply:
@macchiati 
> Wouldn't quite work. Again, we have defined the resolved value of {|3| :number style=percent} to be NOT the same as |3| — since the resolved value includes options. I think what we could say:
>> (either of which can be from the resolved value of a previous declaration).

I think this formulation is actually counterproductive. The _operand_ is not extracted from the _resolved value_. The _operand_ **includes** the accumulated options (which might affect the processing of the _function_ or which might be filtered or ignored).

I agree that the _operands_ in the two expressions in your comments have different _resolved values_, but one of the differences is that `{|3| :number style=percent}`'s _resolved value_ is likely to be an implementation-defined numeric type (such as Java `BigInteger` or `BigDecimal`) and `:math` wouldn't have to parse the literal `|3|` again. 

In any case, my formulation does not say "from" on purpose.

The later "perfect" quote relates to the resolved value _of the `:math` function_, which is the **output** of the function. My clarification moved the word 'resolved' to try to make clear that the options and their values were both resolved. But this is a different context from the _operand_ discussion just above.

- Operand: a naked argument or a _resolved value_ (which is...)
- Resolved value: a resolved (and possibly modified) operand plus resolved options and their resolved values

Oddly, this discussion appears to be about a stock formulation in all of our functions. Can we merge `:math` and then make any changes globally? Can I get your approval on this PR? I filed #937 to track this conversation so we make any agreed changes.

Reply:
Sure.